### PR TITLE
Exposed ForceLightning() to ZScript. And added the ability to change the sound lightning makes.

### PR DESCRIPTION
**Pull request by inkoalawetrust**
This PR exposes the ForceLightning function to ZScript. Allowing for ZScript code to create lightning strikes at will without the use of the Light_ForceLightning line special. This can be used to create custom thunderstorm code that does not rely on the conditions and RNG of the original lightning code. Like the example file attached below.

The PR also adds a LightningSound MAPINFO key, which can be used to define a custom sound for lightning strikes to play on the map. Such as for having different maps use different lightning sounds (e.g one map has nearby thunder, while the next has distant thunder).
It also adds a tempSound parameter to ForceLightning(), to allow thunder to play different sounds on a per-strike basis.

If the LightningSound MAPINFO key isn't used at all, then world/thunder will play as the default, maintaining backwards compatibility. Passing an empty string to the MAPINFO property will make the lightning produce no sound at all. Passing an empty string on ForceLightning() makes it play the maps' lightning sound.

The example mod demonstrates several combinations of lightning behavior. MAP01 has stock lightning, no custom sounds or anything. MAP02 produces silent thunder by using an empty LightningSound property, MAP03 uses the Zombiemans' random sight sounds for thunder. And MAP04 uses an entirely custom lightning implementation that ignores the MAPINFO property for its' own sound, produces thunder more frequently. It will stop for when the player dies and resume when they come back to life. After 20 strikes, the custom thunderstorm thinker stops and brings back the normal thunder behavior, but produces one last explosion sound.

[Lightning example.zip](https://github.com/ZDoom/gzdoom/files/10528889/Lightning.example.zip)

### CODE REVIEW

File: src/playsim/mapthinkers/a_lightning.cpp [hunk 1]
```diff
@@ -48,11 +48,12 @@ IMPLEMENT_CLASS(DLightningThinker, false, false)
 //
 //----------------------------------------------------------------------------
 
-void DLightningThinker::Construct()
```
Comment (CandiceJoy) [inline src/playsim/mapthinkers/a_lightning.cpp:50]: I would highly recommend doing this instead (note that I don't know if the syntax is EXACTLY right, but you get the idea :P)
`inline void DLightningThinker::Construct() : DLightningThinkerConstruct("world/thunder"){}`

File: src/playsim/mapthinkers/a_lightning.h [hunk 2]
```diff
@@ -12,11 +12,11 @@ class DLightningThinker : public DThinker
 	DECLARE_CLASS (DLightningThinker, DThinker);
 public:
 	static const int DEFAULT_STAT = STAT_LIGHTNING;
-	void Construct();
```
Comment (CandiceJoy) [inline src/playsim/mapthinkers/a_lightning.h:15]: Obviously, add definitions here to match earlier suggestions.

File: src/g_levellocals.h [hunk 3]
```diff
@@ -194,7 +194,7 @@ struct FLevelLocals
 	AActor *SpawnMapThing(int index, FMapThing *mt, int position);
 	AActor *SpawnPlayer(FPlayerStart *mthing, int playernum, int flags = 0);
 	void StartLightning();
-	void ForceLightning(int mode);
```
Comment (CandiceJoy) [inline src/g_levellocals.h:197]: `inline void ForceLightning(int mode) { ForceLightning(mode, "world/thunder"); }`

File: src/playsim/mapthinkers/a_lightning.cpp [hunk 4]
```diff
@@ -210,16 +212,18 @@ void DLightningThinker::LightningFlash ()
 //
 //----------------------------------------------------------------------------
 
-void DLightningThinker::ForceLightning (int mode)
```
Comment (CandiceJoy) [inline src/playsim/mapthinkers/a_lightning.cpp:212]: `inline void DLightningThinker::ForceLightning (int mode) { ForceLightning( mode, "world/thunder" ); }`

File: src/playsim/mapthinkers/a_lightning.cpp [hunk 5]
```diff
@@ -286,23 +290,24 @@ void FLevelLocals::StartLightning ()
 //
 //----------------------------------------------------------------------------
 
-void FLevelLocals::ForceLightning (int mode)
```
Comment (CandiceJoy) [inline src/playsim/mapthinkers/a_lightning.cpp:288]: `inline void FLevelLocals::ForceLightning(int mode){ ForceLightning(mode, "world/thunder"); }`

File: src/playsim/mapthinkers/a_lightning.h [hunk 6]
```diff
@@ -12,11 +12,11 @@ class DLightningThinker : public DThinker
...
+	void Construct(FString sound);
 	~DLightningThinker ();
 	void Serialize(FSerializer &arc);
 	void Tick ();
-	void ForceLightning (int mode);
```
Comment (CandiceJoy) [inline src/playsim/mapthinkers/a_lightning.h:19]: Obviously, add definitions here to match earlier suggestions.

**Review by CandiceJoy — CHANGES_REQUESTED** [review]
I would recommend leaving the original DLightningThinker constructor and ForceLightning functions somewhat intact while doing this; specific suggestions attached.

**Comment by inkoalawetrust** [comment]
I don't think the way the constructor is handled is too much of an issue, I basically based it on how DEarthquake does it, and that class also has way more variables than one. The changes to ForceLighting also aren't much of an issue IMO. Before this PR, the line special was the only place that used the function, and I've already changed that to handle the new changes. In fact, I kinda added the current handling of the thunder sound (As opposed to just replacing world/thunder with Level->LightningSound in the S_Sound call), because exposing ForceLightning to ZScript may potentially block adding that sound parameter in down the line.

**Review by coelckers — CHANGES_REQUESTED** [review]
Passing strings by value should be avoided, unless the function needs it to be mutable. Otherwise, pass by reference, this is far more efficient.
Also, make sure that the sound gets restored for old savegames.

**Comment by inkoalawetrust** [comment]
Okay, I've made the fixes. I've also playtested the old savegame fix with a save I made on Hexens' MAP01 in GZDoom 4.7.1, and the fallback to the original hardcoded sound works.

I've also changed how the sound is handled by ForceLightning(). Now the lightning thinkers' sound is a temporary one that is played only once, and can be passed by the function before the thinker goes back to using the maps' LightningSound.

**Comment by inkoalawetrust** [comment]
Well, is this going to be merged ? I've fixed the problem with the lightning sound not working on prior save games.

**Pull request merged**